src/plugins/workspace/public/components/workspace_creator/workspace_permission_setting_panel.tsx

```diff
@@ -227,63 +239,89 @@ export const WorkspacePermissionSettingPanel = ({
   onChange,
   firstRowDeletable,
 }: WorkspacePermissionSettingPanelProps) => {
-  const valueRef = useRef(value);
-  valueRef.current = value;
+  const transferredValue = useMemo(() => {
+    if (!value) {
+      return [];
+    }
+    const result: Array<Partial<WorkspacePermissionSetting>> = [];
+    for (let i = 0; i < value.length; i++) {
+      const valueItem = value[i];
+      if (
+        !valueItem.modes ||
+        !valueItem.type ||
+        (valueItem.type === 'user' && !valueItem.userId) ||
+        (valueItem.type === 'group' && !valueItem.group)
+      ) {
+        result.push(valueItem);
+        continue;
+      }
+      for (const key in optionIdToWorkspacePermissionModesMap) {
+        if (!Object.prototype.hasOwnProperty.call(optionIdToWorkspacePermissionModesMap, key)) {
+          continue;
+        }
+        const modesForCertainPermissionId = optionIdToWorkspacePermissionModesMap[key];
+        if (modesForCertainPermissionId.every((mode) => valueItem.modes?.includes(mode))) {
+          result.push({ ...valueItem, modes: modesForCertainPermissionId });
+        }
+      }
+    }
```

Comment:
The passed permission records were based user & group level. One user or group will contains multi permission modes. Since each permission mode in the UI was single select, we can't display multi permission option in one row. So we need to separate these permission modes into multi rows. Will add some annotation to clarify the logic.